<a href="https://colab.research.google.com/github/johnkinuthiaa/lux-Dev-Data-science-challenges/blob/main/lux_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install pandas

In [ ]:
import pandas as pd
import numpy as np
df =pd.read_csv('/1. Weather_Data.csv')


In [ ]:
# print the first 10 rows of the dataset
df.head(10)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   
5  1/1/2012 5:00    -1.4              -3.3         87                9   
6  1/1/2012 6:00    -1.5              -3.1         89                7   
7  1/1/2012 7:00    -1.4              -3.6         85                7   
8  1/1/2012 8:00    -1.4              -3.6         85                9   
9  1/1/2012 9:00    -1.3              -3.1         88               15   

   Visibility_km  Press_kPa     Weather_Condition  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog  
5            6.4     101.27                   Fog  
6            6.4     101.29                   Fog  
7            8.0     101.26                   Fog  
8            8.0     101.23                   Fog  
9            4.0     101.20                   Fog

In [ ]:
# print the last 10 rows of the dataset
df.tail(10)

Date/Time  Temp_C  Dew_Point_Temp_C  Rel_Hum_%  Wind_Speed_km/h  \
8774  12/31/2012 14:00    -3.4              -5.7         84                6   
8775  12/31/2012 15:00    -2.3              -4.6         84                9   
8776  12/31/2012 16:00    -1.4              -4.0         82               13   
8777  12/31/2012 17:00    -1.1              -3.3         85               19   
8778  12/31/2012 18:00    -1.3              -3.1         88               17   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa Weather_Condition  
8774           11.3     100.57              Snow  
8775            9.7     100.47              Snow  
8776           12.9     100.40              Snow  
8777            9.7     100.30              Snow  
8778            9.7     100.19              Snow  
8779            9.7     100.13              Snow  
8780            9.7     100.03              Snow  
8781            4.8      99.95              Snow  
8782            9.7      99.91              Snow  
8783           11.3      99.89              Snow

In [ ]:
# show the statistics of the dataset
df.describe()

Temp_C  Dew_Point_Temp_C    Rel_Hum_%  Wind_Speed_km/h  \
count  8784.000000       8784.000000  8784.000000      8784.000000   
mean      8.798144          2.555294    67.431694        14.945469   
std      11.687883         10.883072    16.918881         8.688696   
min     -23.300000        -28.500000    18.000000         0.000000   
25%       0.100000         -5.900000    56.000000         9.000000   
50%       9.300000          3.300000    68.000000        13.000000   
75%      18.800000         11.800000    81.000000        20.000000   
max      33.000000         24.400000   100.000000        83.000000   

       Visibility_km    Press_kPa  
count    8784.000000  8784.000000  
mean       27.664447   101.051623  
std        12.622688     0.844005  
min         0.200000    97.520000  
25%        24.100000   100.560000  
50%        25.000000   101.070000  
75%        25.000000   101.590000  
max        48.300000   103.650000



```python
df.describe()
```
This piece of code helps us to get the analysis of the whole dataset.The mean visibility is 0.2 as seen from the response.
It also helps us see themean values for all the other weather conditions and they are as follows:

*   temp_c: 8.798144.
*   Dew point temp_c:2.555294.
*   Rel Hum_%:67.431694.
*   wind speed_km/h:14.945469.
*   visibility_km:27.664447.
*   press_kpa:101.051623.






**CLEANING THE DATASET**

After inspecting the data,we go ahead and clean it by filling some missing and null values


In [ ]:
# rename the weather column
df.rename(columns={'Weather':'Weather_Condition'}, inplace=True)
# rename the wind speed column for easier reading in sql
df.rename(columns={'Wind Speed_km/h':'Wind_Speed_km/h'}, inplace=True)
df.rename(columns={'Dew Point Temp_C':'Dew_Point_Temp_C'}, inplace=True)
df.rename(columns={'Dew Point Temp_C':'Dew_Point_Temp_C'}, inplace=True)
df.rename(columns={'Rel Hum_%':'Rel_Hum_%'}, inplace=True)

In [ ]:
# check whether the column has been renamed successfully
df.head(10)

Date/Time  Temp_C  Dew_Point_Temp_C  Rel_Hum_%  Wind_Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   
5  1/1/2012 5:00    -1.4              -3.3         87                9   
6  1/1/2012 6:00    -1.5              -3.1         89                7   
7  1/1/2012 7:00    -1.4              -3.6         85                7   
8  1/1/2012 8:00    -1.4              -3.6         85                9   
9  1/1/2012 9:00    -1.3              -3.1         88               15   

   Visibility_km  Press_kPa     Weather_Condition  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog  
5            6.4     101.27                   Fog  
6            6.4     101.29                   Fog  
7            8.0     101.26                   Fog  
8            8.0     101.23                   Fog  
9            4.0     101.20                   Fog

In [ ]:
# check for missing values in the dataset
df.isnull().values.any()
# this helps us to see thetotal missing values in all the columns
df.isnull().sum()

Date/Time            0
Temp_C               0
Dew_Point_Temp_C     0
Rel_Hum_%            0
Wind_Speed_km/h      0
Visibility_km        0
Press_kPa            0
Weather_Condition    0
dtype: int64

**Export the new csv to sql**


In [ ]:
df.to_csv('/kaggle/weather_data.csv')



**Running queries on the data using sql**

 Download the new updated csv file to your local storage then load it into a database server application such as dbeaver or xampp so as to run queries on the data so as to get the required information.


1.   To get all the times where the wind speed was exactly 4km/h,we run the query below


```sql
SELECT * FROM `weather_data` WHERE `Wind_Speed_km/h`=4;
```


2.   To get all the times when the wind speed was greater than 24km/hr and visibility was equal to 25 km we run the query below using the logical AND(&&)operator


```sql
<!-- all rows with wind speed above 24 and visibility of 25km -->
SELECT * FROM `weather_data` WHERE (`Wind_Speed_km/h`>24&&`Visibility_km`=25);
```


3.   To get all the instances the weather was clear and the humidity is greator than 50,or visibility is above 40 we use the logical AND(&&) and OR(||):
```sql
SELECT * FROM `weather_data` WHERE (`Weather_Condition`='clear'&&`Rel_Hum_%`>50||`Visibility_km`>40);
```
4.   The number of conditions that innclude snow:
```sql
<!-- this query returns all the times that there was snow -->
SELECT * FROM `weather_data` WHERE `Weather_Condition`='Snow';
```


